84-sarcastic-fringehead/technical_specification.md

```diff
@@ -0,0 +1,742 @@
+     - FIS decrypts the outer token to learn which Data Hub public key to use to decrypt the inner token. The inner token certifies that the request was indeed sent from the given Data Hub.
+     - Inner token, signed by the Data Hub private key, contains the file ID
+   - Request body must contain a payload conforming to the `InterrogationReport` [schema](#interrogationreport)
+   - Returns `204 NO CONTENT`
+   - Description:
+     - FIS finds the matching `FileUnderInterrogation` in its database based on the `file_id` (or raises an error).
+     - If `InterrogationReport.passed` is True, FIS publishes an `InterrogationSuccess` event.
+     - If `InterrogationReport.passed` is False, FIS publishes an `InterrogationFailure` event.
+     - FIS sets both `FileUnderInterrogation.interrogated` and `FileUnderInterrogation.can_remove` to True
+
+#### FIS Configuration
+The FIS needs the following configuration:
+- MongoDbConfig
+- KafkaConfig
+- ApiConfigBase
+- LoggingConfig
+- OpenTelemetryConfig
+- EventSubConfig:
+  - ghga-event-schemas -> `FileUploadEventsConfig`
+  - ghga-event-schemas -> `FileInternallyRegisteredEventsConfig`
+  - ghga-event-schemas -> `InterrogationSuccessEventsConfig`
+  - ghga-event-schemas -> `InterrogationFailureEventsConfig`
+- OutboxSubConfig:
+  - ghga-event-schemas -> `FileUploadEventsConfig`
+- ekss_api_url
+- wkvs_api_url
+
+#### FIS Migrations
+The current FIS implementation has a persisted events collection, `fisPersistedEvents`, that contains previously published `FileInterrogationSuccessEvents` (`type_` is `"file_interrogation_success"`). These events are essentially the same as the events that will, going forward, be published by the UCS upon file archival. Therefore, we *also* need to make sure the FIS can publish to the same outbox topic as the UCS. The required migration for the persisted events collection in FIS essentially needs to convert the `payload` of the stored persisted events into outbox events with the new `FileUpload` [schema](#fileupload). This migration should be reversible. FIS will only publish to the `FileUpload` outbox topic if we *republish* events from the FIS. In other words, this is historical data that we are preserving here in FIS for continuity. The historical `FileUpload` information and the local copies of new `FileUpload` objects must use different DAOs. The former should use an outbox DAO while the latter uses a regular DAO. 
+- **One alternative** is to manually exfiltrate the data to UCS.
+- **A second alternative** is to simply drop the data since the files have already been archived and all relevant information is actually stored by IFRS.
+
+There is one **problem**: IFRS has different object IDs than FIS. If we keep the event data which FIS currently has, we will need to update the information so that the file IDs in FIS are set to the file IDs (object IDs) known to IFRS, using the accession to match. That action would not be reversible, of course.
```

Comment:
The failure mode here would be IFRS loosing information in its internal DB about files that are already archived.

Theoretically we should preserve the information to keep guarantees about what we are able to restore by replaying events, even though personally I'd prefer a clean cutoff as there's a separation between current and new FIS which, in my mind, make them two distinct services and the new one shouldn't be burdened by one off logic permanently in it's code.
Would it make sense to possibly keep this legacy migration separate as a one off script?

On another note, why does this have to be reversible?